spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -0,0 +1,283 @@
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{Node, T, Table}
+
+import scala.reflect.ClassTag
+
+/**
+ * A graph container. Each node can have multiple inputs. The output of the node should be a tensor.
+ * The output tensor can be connected to multiple nodes. So the module in each node can have a
+ * tensor or table input, and should have a tensor output.
+ *
+ * The graph container can have multiple inputs and multiple outputs. If there's one input, the
+ * input data fed to the graph module should be a tensor. If there're multiple inputs, the input
+ * data fed to the graph module should be a table, which is actually an sequence of tensor. The
+ * order of the input tensors should be same with the order of the input nodes. This is also
+ * applied to the gradient from the module in the back propagation.
+ *
+ * If there's one output, the module output is a tensor. If there're multiple outputs, the module
+ * output is a table, which is actually an sequence of tensor. The order of the output tensors is
+ * same with the order of the output modules. This is also applied to the gradient passed to the
+ * module in the back propagation.
+ *
+ * All inputs should be able to connect to outputs through some paths in the graph. It is
+ * allowed that some successors of the inputs node are not connect to outputs. If so, these nodes
+ * will be excluded in the computation.
+ *
+ * @param inputs input nodes
+ * @param outputs output nodes
+ * @tparam T Numeric type. Only support float/double now
+ */
+class Graph[T: ClassTag](inputs : Seq[ModuleNode[T]],
+  outputs : Seq[ModuleNode[T]])(implicit ev: TensorNumeric[T])
+  extends Container[Activity, Activity, T]{
+
+  override def updateOutput(input: Activity): Activity = {
+    var i = 0
+    if (inputs.length == 1) {
+      // A tensor is expected here
+      inputsBP(i) = input.toTensor[T]
+      inputs(i).element.updateOutput(inputsBP(i))
+      i += 1
+    } else {
+      val inputTable = input.toTable
+      require(inputTable.length() == inputs.length,
+        "Input tensor number is not equal to inputs node number")
```

Comment:
changed. But I don't quite understand this as you can see where is the error from in the stack, why need to add Graph.updateOutput?